File: android-gif-drawable/src/androidTest/java/pl/droidsonroids/gif/GifDrawableExceptionTest.java
package pl.droidsonroids.gif;

import android.content.res.Resources;
import androidx.test.platform.app.InstrumentationRegistry;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import pl.droidsonroids.gif.test.R;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

@RunWith(AndroidJUnit4.class)
public class GifDrawableExceptionTest {
	private GifDrawable gifDrawable;

	@Before
	public void setUp() throws Exception {
		final Resources resources = InstrumentationRegistry.getInstrumentation().getContext().getResources();
		gifDrawable = new GifDrawable(resources, R.raw.test);
	}

	@After
	public void tearDown() {
		gifDrawable.recycle();
	}

	@Test
	public void frameIndexOutOfBoundsMessageContainsRange() {
		final int numberOfFrames = gifDrawable.getNumberOfFrames();
		final int invalidFrameIndex = numberOfFrames + 10;
		try {
			gifDrawable.getFrameDuration(invalidFrameIndex);
			fail("Expected IndexOutOfBoundsException to be thrown");
		} catch (IndexOutOfBoundsException exception) {
			assertNotNull(exception.getMessage());
			assertTrue(exception.getMessage().contains(Integer.toString(numberOfFrames)));
		}
	}

	@Test
	public void exceptionThrownWhenPixelsArrayTooSmall() {
		try {
			gifDrawable.getPixels(new int[0]);
			fail("Expected ArrayIndexOutOfBoundsException to be thrown");
		} catch (ArrayIndexOutOfBoundsException exception) {
			// expected
		}
	}

	@Test
	public void exceptionThrownWhenPixelCoordinateXOutOfRange() {
		try {
			gifDrawable.getPixel(gifDrawable.getIntrinsicWidth(), 0);
			fail("Expected IllegalArgumentException to be thrown");
		} catch (IllegalArgumentException exception) {
			// expected
		}
	}

	@Test
	public void exceptionThrownWhenPixelCoordinateYOutOfRange() {
		try {
			gifDrawable.getPixel(0, gifDrawable.getIntrinsicHeight());
			fail("Expected IllegalArgumentException to be thrown");
		} catch (IllegalArgumentException exception) {
			// expected
		}
	}
}

File: android-gif-drawable/src/androidTest/java/pl/droidsonroids/gif/GifDrawableExceptionTest.kt
package pl.droidsonroids.gif

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import org.hamcrest.CoreMatchers
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.ExpectedException
import org.junit.runner.RunWith
import pl.droidsonroids.gif.test.R

@RunWith(AndroidJUnit4::class)
class GifDrawableExceptionTest {

    @get:Rule
    var expectedException = ExpectedException.none()

    private lateinit var gifDrawable: GifDrawable

    @Before
    fun setUp() {
        val resources = InstrumentationRegistry.getInstrumentation().context.resources
        gifDrawable = GifDrawable(resources, R.raw.test)
    }

    @After
    fun tearDown() {
        gifDrawable.recycle()
    }

    @Test
    fun frameIndexOutOfBoundsMessageContainsRange() {
        val numberOfFrames = gifDrawable.numberOfFrames
        val invalidFrameIndex = numberOfFrames + 10

        expectedException.expectMessage(CoreMatchers.containsString(numberOfFrames.toString()))
        expectedException.expect(IndexOutOfBoundsException::class.java)
        gifDrawable.getFrameDuration(invalidFrameIndex)
    }

    @Test
    fun exceptionThrownWhenPixelsArrayTooSmall() {
        expectedException.expect(ArrayIndexOutOfBoundsException::class.java)
        gifDrawable.getPixels(IntArray(0))
    }

    @Test
    fun exceptionThrownWhenPixelCoordinateXOutOfRange() {
        expectedException.expect(IllegalArgumentException::class.java)
        gifDrawable.getPixel(gifDrawable.intrinsicWidth, 0)
    }

    @Test
    fun exceptionThrownWhenPixelCoordinateYOutOfRange() {
        expectedException.expect(IllegalArgumentException::class.java)
        gifDrawable.getPixel(0, gifDrawable.intrinsicHeight)
    }
}